Adding a warning header when a license is about to expire

BigPandaToo opened the pull request:
This implementation will add the warning header 
if the license is going to expire in less than 
{`LICENSE_EXPIRATION_WARNING_PERIOD`} days. 
The messages added:

`Warning: 299 Elasticsearch-8.0.0-###"Your license will expire in [N] days. Contact your administrator or update your license for continued use of features"`

or

`Warning: 299 Elasticsearch-8.0.0-### "Your license expires today. Contact your administrator or update your license for continued use of features"`

If license has expired less than 
{`GRACE_PERIOD_DURATION`} days ago following
 warning is added:

`Warning: 299 Elasticsearch-8.0.0-### "Your license expired on ["EEEE, MMMM dd, yyyy" ]. Contact your administrator or update your license for continued use of features"`

Both {`LICENSE_EXPIRATION_WARNING_PERIOD`} 
and {`GRACE_PERIOD_DURATION`} are currently 7 days.

The message will be added to each request unless 
authentication fails.
Note: with this change all warning headers will be removed 
from a response if authentication fails.

Resolves #60562

Commit:
This change adds a warning header when a license is about to expire

Resolves #60562

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Warning_header

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Warning_header

BigPandaToo commented:
@elasticmachine run elasticsearch-ci/bwc

BigPandaToo commented:
@elasticmachine run elasticsearch-ci/bwc

BigPandaToo commented:
@elasticmachine update branch

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Warning_header

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Warning_header

Commit:
This change adds realm name of the realm used to perform authentication to the responses of _security/oidc/authenticate and _security/oidc/authenticate APIs

Resolves #53161

jkakavas commented on x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:
```diff
@@ -400,9 +404,13 @@ private static boolean isBasic(OperationMode mode) {
         final boolean active;
 
-        Status(OperationMode mode, boolean active) {
+        /** The current expiration date of the license; Long.MAX_VALUE if not avILble yet. */
```
s/avILble/available. Nit: also change the description of the Status class above to add the expiration date

jkakavas commented on x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:
```diff
@@ -508,9 +517,18 @@ public boolean isActive() {
         }
+
+        if(feature.minimumOperationMode.compareTo(OperationMode.BASIC) > 0
+            && now > status.licenseExpiryDate - GRACE_PERIOD_DURATION.getMillis()) {
```
I think we'd need to access this by calling `checkAgainstStatus` instead of accessing the licenseExpiryDate  field directly

jkakavas commented on x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:
```diff
@@ -508,9 +517,18 @@ public boolean isActive() {
             maxEpochAccumulator.accumulate(epochMillisProvider.getAsLong());
         }
+
+        if(feature.minimumOperationMode.compareTo(OperationMode.BASIC) > 0
```
nit: `if (feature....`

jkakavas commented on x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:
```diff
@@ -508,9 +517,18 @@ public boolean isActive() {
+        if(feature.minimumOperationMode.compareTo(OperationMode.BASIC) > 0
+            && now > status.licenseExpiryDate - GRACE_PERIOD_DURATION.getMillis()) {
+            HeaderWarning.addWarning("Your license will expire in [{}] days. " +
+                    "Contact your administrator or update your license for continued use of features",
```
I'm not sure about the wording here but I'll defer to a native speaker and maybe get Fabio's input too? 

Maybe s/update/renew ?  and s/for continued use of/in order to continue using the licensed ?

BigPandaToo replied:
This is the wording Kibana has. I talked to Fabio and he suggested us to be consistant with wording when possible.

jkakavas replied:
SGTM then!

jkakavas reviewed (commented):
I left a few comments. I also think we need to elaborate on the PR description ( it's not clear what `through SECURITY_IP_FILTER`  means ) and add a couple of sentences on why we add the check on `checkFeature` and more especially why we think this is enough for solving the issue this PR is meant to. 

We also need some tests that ensure we are indeed adding the warning header under the circumstances we expect it to be added

elasticmachine commented:
Pinging @elastic/es-security (Team:Security)

Commit:
Adding doc for the new API introduced by #64517 - /_security/saml/metadata/{realm}

Related to #49018

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Warning_header

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:
```diff
@@ -508,9 +517,17 @@ public boolean isActive() {
         }
+
+        if (feature.minimumOperationMode.compareTo(OperationMode.BASIC) > 0 && isLicenseExpiring(now)) {
+            HeaderWarning.addWarning("Your license will expire in [{}] days. " +
```
If we're in the grace period, shouldn't this say `has expired` ?

BigPandaToo replied:
I'll double-check with Fabio. This is the only message Kibana has and the idea was to be consistent when possible.

bytebilly replied:
I don't think we have a grace period anymore. AFAIK we have warnings before the expiration date, and no service after that. Could you please double check what `isLicenseExpiring()` assumes?

tvernum replied:
We still have a grace period. We may not in the future, but we do now.

bytebilly replied:
_(sorry I cannot find the description of the full approach, so it may be already this)_

What about having two separate checks with two separate messages?

- 7 days before expiration — expiration: "will expire in N days"
- expiration — end of grace period: "has expired N days ago"

BigPandaToo replied:
I changed the behaviour to have 3 different messages:
 - 7 days before expiration — expiration: "will expire in N days"
- <1 days before expiration: "expires today"
- > expiration: "has expired N days ago"
If grace period is over this code should be irrelevant as the license is not active.

BigPandaToo replied:
Also, addad a test

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:
```diff
@@ -627,6 +644,22 @@ public boolean isAllowedByLicense(OperationMode minimumMode, boolean needActive)
     }
 
+    /**
+     * Test whether current license expires in less than {@code GRACE_PERIOD_DURATION}.
```
Why is this dependent on the `GRACE_PERIOD` ? It seems to be a co-incidence that they're both 7days, but they shouldn't be tied to one another.

BigPandaToo replied:
Good point, I'll add new const

tvernum reviewed (changes requested):
It doesn't look like there are any tests for the new warning.

I think we need some (at the very least to verify that the time calculations are handled correctly).

Commit:
Adding a warning header when a license is about to expire

Resolves #60562

tvernum commented:
It looks like we return the warning header even if authentication fails. I don't think we should do that.

bytebilly reviewed (commented):
@BigPandaToo sorry I'm not confident I can get the exact behavior of this feature looking at the code only.
Could you please briefly describe high-level actions?
I'd want to get answers to questions like:
- when the message is sent (time based)
- where the message is sent (every request, requests for paid features only, etc)
- what is sent (message content, code, etc)

That would be very helpful to provide some feedback.
Thanks!

jkakavas commented on x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:
```diff
@@ -627,6 +650,22 @@ public boolean isAllowedByLicense(OperationMode minimumMode, boolean needActive)
+     *
+     * @return true if current license expires in less than {@code LICENSE_EXPIRATION_WARNING_PERIOD}, otherwise false
+     */
+    public boolean isLicenseExpiring(long now) {
```
Given what this does, would it make sense to name it `shouldWarnAboutLicense()` ?

jkakavas commented on x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/license/LicensingTests.java:
```diff
@@ -235,8 +293,42 @@ private void enableLicensing(License.OperationMode operationMode) throws Excepti
+        // overwritten by some other cluster activity and the node throws an exception while we
+        // wait for things to stabilize!
+        assertBusy(() -> {
+            // first update the license so we can execute monitoring actions
```
Do we have to do the double update dance here? `setLicensingExpirationDate` is the only thing that calls update with anything other than Long.MAX_VALUE, what cluster state could trigger the license state (Expiration date) to change ?

BigPandaToo replied:
Probably, not. Removed the second update, left hte assertBusy though.

jkakavas commented on x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:
```diff
@@ -505,12 +516,24 @@ public boolean isActive() {
+
+        if (feature.minimumOperationMode.compareTo(OperationMode.BASIC) > 0 && isLicenseExpiring(now)) {
+            final long days = TimeUnit.MILLISECONDS.toDays(status.licenseExpiryDate - now);
+            final String expiryMessage = days == 0? "expires today":
```
```suggestion
            final String expiryMessage = days == 0 ? "expires today" :
```

jkakavas commented on x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:
```diff
@@ -505,12 +516,24 @@ public boolean isActive() {
+        if (feature.minimumOperationMode.compareTo(OperationMode.BASIC) > 0 && isLicenseExpiring(now)) {
+            final long days = TimeUnit.MILLISECONDS.toDays(status.licenseExpiryDate - now);
+            final String expiryMessage = days == 0? "expires today":
+                (days > 0? String.format(Locale.ROOT, "will expire in [%d] days", days):
```
```suggestion
                (days > 0 ? String.format(Locale.ROOT, "will expire in [%d] days", days):
```

jkakavas reviewed (commented):
We still need a proper PR description to accompany the commit once this is merged.

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Warning_header

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Warning_header

BigPandaToo commented:
> It looks like we return the warning header even if authentication fails. I don't think we should do that.

Changing the approach. Moving the warning adding block to SecurityRestFilter. There are pros and cons though:
pros:
 - we can avoid adding the headers when authentication fails
 - the message is sent with all requests when security is enabled
cons:
 - no message when security is not enabled

BigPandaToo commented:
> @BigPandaToo sorry I'm not confident I can get the exact behavior of this feature looking at the code only.
> Could you please briefly describe high-level actions?
> I'd want to get answers to questions like:
> 
> * when the message is sent (time based)
> * where the message is sent (every request, requests for paid features only, etc)
> * what is sent (message content, code, etc)
> 
> That would be very helpful to provide some feedback.
> Thanks!

@bytebilly 
I have changed the code as the previous approach didn't allow avoiding messaging if authentication failed. Please, see the new description.

Commit:
Addressing the PR feedback

bytebilly reviewed (approved):
The high-level approach LGTM 👍

tvernum commented:
> no message when security is not enabled

That sounds like a pretty big downside.
I think we can do something like this -> https://github.com/tvernum/elasticsearch/commit/d8f36cf52f8 <- and then stick with the original version that relied on `XPackLicenseState`.

It would have the effect of removing all warning headers, but I think that's actually preferred, (see also: https://github.com/elastic/elasticsearch/pull/55774)

@tbrooks8, since you seem to be the primary maintainer of `DefaultRestChannel` do you think the commit I linked to just above is an appropriate way to prevent `Warning` headers from being sent on authentication failure responses.

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Warning_header

droberts195 commented on x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:
```diff
@@ -507,12 +513,15 @@ public boolean isActive() {
     /**
      * Checks whether the given feature is allowed, tracking the last usage time.
      */
+    @SuppressForbidden(reason = "Argument to Math.abs() is definitely not Long.MIN_VALUE")
```
Not necessary?  This method doesn't appear to use `Math.abs()`.

BigPandaToo replied:
Leftover from the previous iteration, but will come back handy with the next one )

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Warning_header

Commit:
Switching back to adding the header during featureCheck to allow
warnings when authentication is disabled as well. Adding filterHeader
implementation to SecurityRestFilter exception handling to remove all
the warnings if authentication fails.

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Warning_header

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:
```diff
@@ -507,12 +518,26 @@ public boolean isActive() {
+            final long days = TimeUnit.MILLISECONDS.toDays(licenseExpirationDate - now);
+            final String expiryMessage = days == 0? "expires today":
+                (days > 0? String.format(Locale.ROOT, "will expire in [%d] days", days):
+                    String.format(Locale.ROOT, "has expired [%d] days ago", Math.abs(days)));
```
For expired licenses, I think it's preferable to just print out the date that is expired, rather than calculate how lonh ago that was.
That's what the log warning does and it seems neater for this case.

> LICENSE [EXPIRED] ON [SUNDAY, NOVEMBER 29, 2020]

The number of days makes sense for "will expire" because it creates a sense of urgency, but once it's expired I think the date is a better option.

BigPandaToo replied:
Changed the wording to "license expired on [EEEE, MMMM DD, YYYY]". For the sace of keeping the headers relatively short kept the rest of the message the same.

tvernum commented on x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:
```diff
@@ -507,12 +518,26 @@ public boolean isActive() {
             maxEpochAccumulator.accumulate(epochMillisProvider.getAsLong());
         }
+
+        if (feature.minimumOperationMode.compareTo(OperationMode.BASIC) > 0 &&
+            now >  licenseExpirationDate - LICENSE_EXPIRATION_WARNING_PERIOD.getMillis()) {
+            final long days = TimeUnit.MILLISECONDS.toDays(licenseExpirationDate - now);
+            final String expiryMessage = days == 0? "expires today":
+                (days > 0? String.format(Locale.ROOT, "will expire in [%d] days", days):
+                    String.format(Locale.ROOT, "has expired [%d] days ago", Math.abs(days)));
```
I think we should be doing the first round of calculations in millis, not in days otherwise we potentially says "expires today" _after_ the expiry date (because it's less than 1 day expired)

```suggestion
    
        if (feature.minimumOperationMode.compareTo(OperationMode.BASIC) > 0) {
            final long expiryMillis = licenseExpirationDate - now;        
            if (expiryMillis < LICENSE_EXPIRATION_WARNING_PERIOD.getMillis()) {
                final String expiryMessage;
                if (expiryMillis <= 0) {
                    expiryMessage = "expired on [" + DATE_FORMATTER.formatMillis(expirationMillis) + "]";
                } else {
                    final long days = TimeUnit.MILLISECONDS.toDays(expiryMillis);
                    expiryMessage = days == 0 ? "expires today": 
                       String.format(Locale.ROOT, "will expire in [%d] days", days);
                }
```

BigPandaToo replied:
Done

tvernum commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java:
```diff
@@ -100,6 +102,14 @@ private void handleException(String actionType, RestRequest request, RestChannel
+                    }
+                    return headers;
+                }
+
```
Do we have a test for this anywhere?

BigPandaToo replied:
Added a test to verify Warnings are removed if authN fails.

Commit:
Changing the wording for "expired" message to be consistent with the log
 messages; changing "today" calculation; adding a test case for failing
 authN to make sure we remove the warning header

tvernum reviewed (approved):
LGTM, but I'd like to hear from @tbrooks8 before we merge.
(and if we can switch to `assertThat`, that would be good too)

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Warning_header

Commit:
Small changes in the way we verify header in tests

Tim-Brooks reviewed (approved):
LGTM

jkakavas commented on x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/rest/SecurityRestFilter.java:
```diff
@@ -100,6 +102,14 @@ private void handleException(String actionType, RestRequest request, RestChannel
 
+                @Override
+                public Map<String, List<String>> filterHeaders(Map<String, List<String>> headers) {
+                    if (headers.containsKey("Warning")) {
```
Do we want to remove _all_ warning headers on authentication failure?

BigPandaToo replied:
Yes, it is intentional (see Tim's comment above). If Authentication fails it does make sense to not reveal any additional information

jkakavas commented on x-pack/plugin/core/src/test/java/org/elasticsearch/test/SecuritySettingsSourceField.java:
```diff
@@ -10,6 +10,7 @@
 public final class SecuritySettingsSourceField {
     public static final SecureString TEST_PASSWORD_SECURE_STRING = new SecureString("x-pack-test-password".toCharArray());
     public static final String TEST_PASSWORD = "x-pack-test-password";
+    public static final String TEST_INVALID_PASSWORD = "invalid-test-password";
```
I think you can use a randomAlphaNumericOfLength , or a permutation of `TEST_PASSWORD` plus some random string for whenever we need this, other than defining it here. That's just personal preference, no strong views, feel free to keep as is if you prefer it

jkakavas commented on x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/license/LicensingTests.java:
```diff
@@ -190,6 +202,82 @@ public void testNodeJoinWithoutSecurityExplicitlyEnabled() throws Exception {
+
+        List<String> beforeWarningHeaders = getWarningHeaders(response.getHeaders());
+
+        assertTrue(beforeWarningHeaders.isEmpty());
```
use assertThat ?

jkakavas commented on x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/license/LicensingTests.java:
```diff
@@ -190,6 +202,82 @@ public void testNodeJoinWithoutSecurityExplicitlyEnabled() throws Exception {
+        options.addHeader("Authorization", basicAuthHeaderValue(SecuritySettingsSource.TEST_USER_NAME,
+            new SecureString(SecuritySettingsSourceField.TEST_PASSWORD.toCharArray())));
+        request.setOptions(options);
+
```
nit: Do we need all the empty lines?

BigPandaToo replied:
Yes, it is intentional (see Tim's comment above). If Authentication fails it does make sense to not reveal any additional information

jkakavas commented on x-pack/plugin/core/src/main/java/org/elasticsearch/license/XPackLicenseState.java:
```diff
@@ -509,12 +520,26 @@ public boolean isActive() {
+
+        if (feature.minimumOperationMode.compareTo(OperationMode.BASIC) > 0 &&
+            LICENSE_EXPIRATION_WARNING_PERIOD.getMillis() > diff) {
+            final long days = TimeUnit.MILLISECONDS.toDays(diff);
```
I Guess we are fine with the message in warnigns and subtleties of `toDays` and being close to midnight in a given timezone , right? I think we are, just asking for completeness.

BigPandaToo replied:
I believe we are :)

jkakavas commented on x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/license/LicensingTests.java:
```diff
@@ -190,6 +202,82 @@ public void testNodeJoinWithoutSecurityExplicitlyEnabled() throws Exception {
+            List<String> afterWarningHeaders= getWarningHeaders(e.getResponse().getHeaders());
+            assertThat(afterWarningHeaders, Matchers.hasSize(0));
+        }
+        assertTrue(headers != null && headers.length == 3);
```
use assertThat ?

jkakavas commented on x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/license/LicensingTests.java:
```diff
@@ -190,6 +202,82 @@ public void testNodeJoinWithoutSecurityExplicitlyEnabled() throws Exception {
+            "Contact your administrator or update your license for continued use of features"));
+    }
+
+    public void testWarningHeaderAuthenticationFailed() throws Exception {
```
nit suggestion : testNoWarningHeaderWhenAuthenticationFailed

jkakavas reviewed (commented):
Just some nits suggestions and a question regarding if removing _all_ warning headers on authentication failure is intentional

BigPandaToo commented:
> Just some nits suggestions and a question regarding if removing _all_ warning headers on authentication failure is intentional

@jkakavas Yes, it is intentional (see Tim's comment above). If Authentication fails it does make sense to not reveal any additional information

Commit:
Nit changes

jkakavas reviewed (approved):
LGTM

BigPandaToo commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into Warning_header
